In [4]:
import json

with open("./fall2020/fall2020(jan2).json", "r") as read_file:
    vergil = json.load(read_file)
    read_file.close() 
print(len(vergil))

with open("./spring2021/spring2021(jan2).json", "r") as read_file:
    vergil = json.load(read_file)
    read_file.close()   
print(len(vergil))

with open("./spring2020/spring2020(july20).json", "r") as read_file:
    vergil = json.load(read_file)
    read_file.close()
print(len(vergil))

with open("./fall2020/fall2020ugradData.json", "r") as read_file:
    vergil = json.load(read_file)
    read_file.close() 
print(len(vergil))

with open("./spring2021/spring2021ugradData.json", "r") as read_file:
    vergil = json.load(read_file)
    read_file.close()   
print(len(vergil))

with open("./spring2020/spring2020ugradData.json", "r") as read_file:
    vergil = json.load(read_file)
    read_file.close()
print(len(vergil))

4383
3902
4046
1600
1587
1765


## Reformat original json files
### Missing data:
- spring 2020:
    - ugrad course: 2060
    - ugrad courses with time info: 1765
- fall 2020
    - ugrad course: 1918
    - ugrad courses with time info: 1600
- spring 2021
    - ugrad course: 1860
    - ugrad courses with time info: 1587

In [4]:
import json

with open("./spring2020/spring2020(july20).json", "r") as read_file:
    vergil = json.load(read_file)
    read_file.close()

#put all 0000-4000 level courses in 'undergrad' array
undergrad = []
for course in vergil:
    courseId = course["course"]["course_identifier"]
    
    if len(courseId) == 9 or len(courseId) == 10:
        if courseId[-4].isdigit():
            if int(courseId[-4]) <= 4:
                #print(courseId)
                undergrad.append(course["course"])
                
#print(len(undergrad)) 
#print(undergrad[2:3])

#extract each course section's time and max enrollment
alltimeDicts = {}
types = {}
for course in undergrad:
    courseId = course["course_identifier"]
    
    sections = course["classes"]["class"]
    times = []
    info = {}
    
    
    for section in sections:
        timesListed = section["days_times"]
        size = section["enrollment"]["max"]
        classtype = section["type"]["type_code"]
        classtypename = section["type"]["name"]
        
        #print(classtype, classtypename)
        if classtypename not in types:
            types[classtypename] = 1
        else:
            types[classtypename] += 1
        
        for day in timesListed:
            startT = day["mil_time_from"]
            endT = day["mil_time_to"]
            time = [startT, endT]
            if startT and endT:
                times.append(time)
                if classtypename not in types:
                    types.append(classtypename)
    
    
    if times:
        info.update({"size": size})
        info.update({"times": times})
        info.update({"type": classtypename})
        
    if info:
        alltimeDicts.update({courseId: info})

#for type in types:
    #print(type, types[type])

#print(len(alltimeDicts))
#print(alltimeDicts)

#write to json

#with open("spring2021ugradData.json","w") as write:
    #json.dump(alltimeDicts, write, indent = 4)

Lecture 1421
Seminar 537
Recitation 115
Independent 748
Research Sem 33
Laboratory 161
Language 354
Conversation 11
Tutorial 64
Discussion 262
Field Work 12
Colloquia 37
Internship 5
Dummy Course 23
Studio 63
Residency 4
Registration 4
Workshop 88
Film Screeni 1
Physical Edu 106
Clinic 11
Rehearsal 2
Practica 18
Oral/Aural P 1


## scrap: convert time zones to .json

In [17]:
from datetime import datetime, timezone
import pytz
from parse import *
import json

with open("./spring2020/spring2020ugradData.json", "r") as read_file:
    ugrad = json.load(read_file)
    read_file.close()
    
#for tz in pytz.all_timezones:
    #print(tz)

#set standard date as 01-01-2020 
#not during daylight savings, so 23 hrs between nyc and beijing instead of 24
standard = "01-01-2020 "

fmt = "%H:%M"

#timezone variables
ny = pytz.timezone('America/New_York')
bj = pytz.timezone('Asia/Shanghai')
la = pytz.timezone('America/Los_Angeles')
london =pytz.timezone('Europe/London')
moscow =pytz.timezone('Europe/Moscow')
vancouver =pytz.timezone('Etc/GMT+8')
mexico =pytz.timezone('Mexico/General')
sao_paolo =pytz.timezone('America/Sao_Paulo')
athens =pytz.timezone('Europe/Athens')
baghdad =pytz.timezone('Asia/Baghdad')
bangkok =pytz.timezone('Asia/Bangkok')
syndey =pytz.timezone('Australia/Sydney')

def convert(targetZone, timeStr):
    dt = datetime.strptime(timeStr, "%d-%m-%Y %H:%M:%S")
    dt = ny.localize(dt)
    dtconverted = dt.astimezone(targetZone)
    return dtconverted.strftime(fmt)

arr = ugrad

'''
with open('spring2021_syndey_endTimes.csv', mode='w', newline='') as file:
    writer = csv.writer(file)
    for course in arr:
        for time in arr[course]["times"]:
            start = time[0]
            start = standard + start + ":00"
            time[0] = convert(bj, start)

            end = time[1]
            end = standard + end + ":00"
            time[1] = convert(bj, end)
            writer.writerow([time[1]])

for course in arr:
    for time in arr[course]["times"]:
        start = time[0]
        start = standard + start + ":00"
        time[0] = convert(la, start)

        end = time[1]
        end = standard + end + ":00"
        time[1] = convert(la, end)
        
with open("spring2021ugrad_la.json","w") as write:
    json.dump(arr, write, indent = 4)

with open("spring2021ugrad_syndey.json","w") as write:
    json.dump(arr, write, indent = 4)
    
print(bjArr)
print(bjStartArr)
'''
for course in arr:
        for time in arr[course]["times"]:
            start = time[0]
            start = standard + start + ":00"
            time[0] = convert(bj, start)

            end = time[1]
            end = standard + end + ":00"
            time[1] = convert(bj, end)
with open("spring2020ugrad_bj.json","w") as write:
    json.dump(arr, write, indent = 4)


## Check class sizes

In [7]:
import json
import statistics as stats

with open("./fall2020/fall2020ugradData.json", "r") as read_file:
    courses = json.load(read_file)
    read_file.close()
    
size = []    
nullCount = 0
for course in courses:
    if not courses[course]["size"]:
        #print(course)
        nullCount+=1
    if not courses[course]["times"][0]:
        print(course)
    else:
        if courses[course]["type"] == "Studio":
            #print(course , courses[course]["size"])
            size.append(int(courses[course]["size"]))
        
print(stats.mean(size))

230.94871794871796


## scrap

In [8]:
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from matplotlib.collections import PolyCollection
from collections import Counter
import csv


with open("./spring2021/spring2021ugrad_bj.json", "r") as read_file:
    courses = json.load(read_file)
    read_file.close()

endTimes = []
for course in courses:
    for time in courses[course]["times"]:
        endTime = time[1]
        #print(course,endTime)
        endTimes.append(endTime)

with open('spring2021_bj_endTimes.csv', mode='w', newline='') as file:
    writer = csv.writer(file)
    for course in courses:
        for time in courses[course]["times"]:
            endTime = time[1]
            writer.writerow(endTime)
 
print(len(endTimes))
freq = Counter(endTimes)

#for k,v in freq.items():
    #print(k,v)

'''
with open('spring2020_freq.csv', mode='w', newline='') as file:
    writer = csv.writer(file)
    for k,v in freq.items():
        writer.writerow([k,v])
'''

FileNotFoundError: [Errno 2] No such file or directory: './spring2021/spring2021ugrad_bj.json'

## Convert course end times in a single .csv

In [193]:
import csv
from datetime import datetime, timezone
import pytz
from parse import *

#set standard date as 01-01-2020 
#not during daylight savings, so 23 hrs between nyc and beijing instead of 24
standard = "01-01-2020 "

fmt = "%H:%M"

#timezone variables
ny = pytz.timezone('America/New_York')
bj = pytz.timezone('Asia/Shanghai')
la = pytz.timezone('America/Los_Angeles')
london =pytz.timezone('Europe/London')
moscow =pytz.timezone('Europe/Moscow')
vancouver =pytz.timezone('Etc/GMT+8')
mexico =pytz.timezone('Mexico/General')
sao_paolo =pytz.timezone('America/Sao_Paulo')
athens =pytz.timezone('Europe/Athens')
baghdad =pytz.timezone('Asia/Baghdad')
bangkok =pytz.timezone('Asia/Bangkok')
syndey =pytz.timezone('Australia/Sydney')


with open("C:/Users/Administrator/Desktop/spec/vergil_wrangle/fall2020/fall2020_endTimes.csv", newline='') as file:
    reader = csv.reader(file)
    
    with open("C:/Users/Administrator/Desktop/spec/vergil_wrangle/fall2020_endTimes.csv", 'w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(["nyc", "bj", "la", "london", "moscow", "vancouver", "mexico", "sao_paolo", "athens", "baghdad", "bangkok", "syndey"])
        for row in reader:
            nyc = standard + row[0] + ":00"

            Tnyc = convert(ny, nyc)
            Tbj = convert(bj, nyc)
            Tla = convert(la, nyc)
            Tlondon = convert(london, nyc)
            Tmoscow = convert(moscow, nyc)
            Tvancouver = convert(vancouver, nyc)
            Tmexico = convert(mexico, nyc)
            Tsao_paolo = convert(sao_paolo, nyc)
            Tathens = convert(athens, nyc)
            Tbaghdad = convert(baghdad, nyc)
            Tbangkok = convert(bangkok, nyc)
            Tsyndey = convert(syndey, nyc)

            writer.writerow([Tnyc, Tbj, Tla, Tlondon, Tmoscow, Tvancouver, Tmexico, Tsao_paolo, Tathens, Tbaghdad, Tbangkok, Tsyndey ])


def convert(targetZone, timeStr):
    dt = datetime.strptime(timeStr, "%d-%m-%Y %H:%M:%S")
    dt = ny.localize(dt)
    dtconverted = dt.astimezone(targetZone)
    return dtconverted.strftime(fmt)

## Convert course to endtimes in .csv for each class type

In [21]:
import csv
from datetime import datetime, timezone
import pytz
from parse import *

lecture_types = {"Lecture"}
seminar_types = {"Seminar", "Discussion", "Recitation", "Language"}
lab_types = {"Laboratory"}

#set standard date as 01-01-2020 
#not during daylight savings, so 23 hrs between nyc and beijing instead of 24
standard = "01-01-2020 "

fmt = "%H:%M"

#timezone variables
ny = pytz.timezone('America/New_York')
bj = pytz.timezone('Asia/Shanghai')
la = pytz.timezone('America/Los_Angeles')
london =pytz.timezone('Europe/London')
moscow =pytz.timezone('Europe/Moscow')
vancouver =pytz.timezone('Etc/GMT+8')
mexico =pytz.timezone('Mexico/General')
sao_paolo =pytz.timezone('America/Sao_Paulo')
athens =pytz.timezone('Europe/Athens')
baghdad =pytz.timezone('Asia/Baghdad')
bangkok =pytz.timezone('Asia/Bangkok')
syndey =pytz.timezone('Australia/Sydney')

with open("./spring2021/spring2021ugradData.json", "r") as read_file:
    ugrad = json.load(read_file)
    read_file.close()

with open("./spring2021_endTimes_lecture.csv", 'w', newline='') as file1, \
open("./spring2021_endTimes_seminar.csv", 'w', newline='') as file2, \
open("./spring2021_endTimes_lab.csv", 'w', newline='') as file3:
    writer1 = csv.writer(file1)
    writer2 = csv.writer(file2)
    writer3 = csv.writer(file3)
    writer1.writerow(["nyc", "bj", "la", "london", "moscow", "vancouver", "mexico", "sao_paolo", "athens", "baghdad", "bangkok", "syndey"])
    writer2.writerow(["nyc", "bj", "la", "london", "moscow", "vancouver", "mexico", "sao_paolo", "athens", "baghdad", "bangkok", "syndey"])
    writer3.writerow(["nyc", "bj", "la", "london", "moscow", "vancouver", "mexico", "sao_paolo", "athens", "baghdad", "bangkok", "syndey"])
    for course in ugrad:
        if ugrad[course]["type"] in lecture_types:
            writer = writer1
        if ugrad[course]["type"] in seminar_types:
            writer = writer2
        if ugrad[course]["type"] in lab_types:
            writer = writer3
        if writer:
            for time in ugrad[course]["times"]:
                end = time[1]
                nyc = standard + end + ":00"

                Tnyc = convert(ny, nyc)
                Tbj = convert(bj, nyc)
                Tla = convert(la, nyc)
                Tlondon = convert(london, nyc)
                Tmoscow = convert(moscow, nyc)
                Tvancouver = convert(vancouver, nyc)
                Tmexico = convert(mexico, nyc)
                Tsao_paolo = convert(sao_paolo, nyc)
                Tathens = convert(athens, nyc)
                Tbaghdad = convert(baghdad, nyc)
                Tbangkok = convert(bangkok, nyc)
                Tsyndey = convert(syndey, nyc)

                writer.writerow([Tnyc, Tbj, Tla, Tlondon, Tmoscow, Tvancouver, Tmexico, Tsao_paolo, Tathens, Tbaghdad, Tbangkok, Tsyndey ])


def convert(targetZone, timeStr):
    dt = datetime.strptime(timeStr, "%d-%m-%Y %H:%M:%S")
    dt = ny.localize(dt)
    dtconverted = dt.astimezone(targetZone)
    return dtconverted.strftime(fmt)
    